docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,960 @@
+# Pageserver generation numbers for safe failover & tenant migration
+
+## Summary
+
+A scheme of logical "generation numbers" for pageservers and their attachments is proposed, along with
+changes to the remote storage format to include these generation numbers in S3 keys.
+
+Using the control plane as the issuer of these generation numbers enables strong anti-split-brain
+properties in the pageserver cluster without implementing a consensus mechanism directly
+in the pageservers.
+
+## Motivation
+
+Currently, the pageserver's remote storage format does not provide a mechanism for addressing
+split brain conditions that may happen when replacing a node during failover or when migrating
+a tenant from one pageserver to another. From a remote storage perspective, a split brain condition
+occurs whenever two nodes both think they have the same tenant attached, and both can write to S3. This
+can happen in the case of a network partition, pathologically long delays (e.g. suspended VM), or software
+bugs.
+
+This blocks robust implementation of failover from unresponsive pageservers, due to the risk that
+the unresponsive pageserver is still writing to S3.
+
+### Prior art
+
+- 020-pageserver-s3-coordination.md
+- 023-the-state-of-pageserver-tenant-relocation.md
+- https://www.notion.so/neondatabase/Proposal-Pageserver-MVCC-S3-Storage-8a424c0c7ec5459e89d3e3f00e87657c
```

Comment:
I archived this Notion page to a PR to this repo: https://github.com/neondatabase/neon/pull/5040

Let's get it merged, then update this link.

Reply:
Updated in f545231bcb92fba8182d368c324b3f108f476fbc